### PR TITLE
[SPARK-34819][SQL] MapType supports orderable semantics

### PR DESCRIPTION
**Pull request by WangGuangxin**
### What changes were proposed in this pull request?
Currently MapType doesn't support orderable semantics, while it's supported in Hive/Presto. This makes it hard to migrate from Hive to SparkSQL if user have groupby/orderby map type in their sql.

### Why are the changes needed?
Generally,  we compare two maps by the following steps:
1. If the size of two maps are not equal, compare them by size.
2. Otherwise, sort each map entry by map key, then compare two map entries one by one, first compare by key, then value.

We have to specially handle this in grouping/join/window because Spark SQL turns grouping/join/window partition keys into binary `UnsafeRow` and compare the binary data directly instead of using MapType's ordering. In this case, we have to insert a `SortMapKey` expression to sort map entry by key. This is very similiar to `NormalizeFloatingNumbers` 

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Add more UTs

### CODE REVIEW

**Comment by WangGuangxin** [comment]
@hvanhovell  @cloud-fan @maropu Could you please help review this?

**Review by c21 — COMMENTED** [review]
Is it a re-proposal for https://issues.apache.org/jira/browse/SPARK-18134 ? I thought the decision made on that JIRA is we won't support this. But we also encountered this when migrating from Hive to Spark. We worked around this by adding a logical plan rule to covert map to sorted array if needed.

**Comment by hvanhovell** [comment]
@c21 we should support this. I just ran out of time when I was working on it.

**Comment by hvanhovell** [comment]
Ok to test

**Comment by maropu** [comment]
Link: #15970 and #19330

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala [hunk 1]
```diff
@@ -4140,6 +4142,136 @@ class SQLQuerySuite extends QueryTest with SharedSparkSession with AdaptiveSpark
...
+                Row(Map("c" -> 3, "b" -> 2, "a" -> 1)) :: Nil
+            )
+          }
+        }
+    }
+  }
+
+  test("SPARK-34819: MapType has nesting complex type supports orderable semantics") {
+    Seq(CodegenObjectFactoryMode.CODEGEN_ONLY.toString,
+      CodegenObjectFactoryMode.NO_CODEGEN.toString).foreach {
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala:4239]: Could you move the two tests into `SQLQueryTestSuite`? You can use the `CONFIG_DIM` directive there: 
https://github.com/apache/spark/blob/master/sql/core/src/test/resources/sql-tests/inputs/postgreSQL/join.sql#L18-L20

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NormalizeMapType.scala [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+  private lazy val keyOrdering: Ordering[Any] = TypeUtils.getInterpretedOrdering(keyType)
+  private lazy val mapBuilder = new ArrayBasedMapBuilder(keyType, valueType)
+
+  override def inputTypes: Seq[AbstractDataType] = Seq(MapType)
+
+  override def dataType: DataType = child.dataType
+
+  override def nullSafeEval(input: Any): Any = {
+    val childMap = input.asInstanceOf[MapData]
+    val keys = childMap.keyArray()
```
[the remark this answers is not in the document]
Comment (WangGuangxin) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NormalizeMapType.scala:85]: Seems that I missed this case. I'll fix it

**Comment by maropu** [comment]
Any update?

**Comment by maropu** [comment]
@WangGuangxin If you cannot keep working on it, is it okay that I take this over?

**Comment by WangGuangxin** [comment]
> @WangGuangxin If you cannot keep working on it, is it okay that I take this over?

Sure, I'm stuck with something else, you can take this over if you have time. Thanks

**Pull request closed without merging**